containers/source-build/source_build.py

```diff
@@ -0,0 +1,452 @@
+
+            gathered = True
+
+            # FIXME: perhaps the dependency archive (the Cachi2) might be handled differently
+            run(["tar", "xvf", extra_src_archive], check=True, cwd=dest_dir)
+            os.unlink(f"{dest_dir}/{extra_src_archive}")
+
+        else:
+            log.warning("No known operation happened on layer %s", digest)
+
+    return gathered
+
+
+def gather_prefetched_sources(
+    work_dir: str, cachi2_dir: str, sib_dirs: SourceImageBuildDirectories
+) -> bool:
+    log = logging.getLogger("source-build.prefetched-sources")
+    gathered = False
+
+    # Guess if hermetic build is enabled
+    # NOTE: this guess does depend on how cachi2 runs inside prefetch-dependencies task.
+    cachi2_output_dir = f"{cachi2_dir}/output"
+
+    if not os.path.isdir(cachi2_output_dir):
+        log.info("Cannot find cachi2 output directory at %s", cachi2_output_dir)
+        return gathered
+
+    def _find_prefetch_source_archives():
+        for root, dirs, files in os.walk(cachi2_output_dir):
+            for filename in files:
+                is_ext = filename.endswith
+                # .zip for gomod, .tgz for pip, .tar.gz for npm
+                if is_ext(".zip") or is_ext(".tgz") or is_ext(".tar.gz"):
+                    yield root, filename
+
+    source_counter = itertools.count()
+    prepared_sources_dir = create_dir(work_dir, "prefetched_sources")
+    relative_to = os.path.relpath
+
+    for root, filename in _find_prefetch_source_archives():
+        src_dir = f"src-{next(source_counter)}"
+        copy_dest_dir = f"{prepared_sources_dir}/{src_dir}/{relative_to(root, cachi2_output_dir)}"
+        os.makedirs(copy_dest_dir)
+
+        src = f"{root}/{filename}"
+        dest = f"{copy_dest_dir}/{filename}"
+        log.debug("copy prefetched source %s to %s", src, dest)
+        shutil.copy(src, dest)
+        sib_dirs.extra_src_dirs.append(f"{prepared_sources_dir}/{src_dir}")
+
+    gathered = next(source_counter) > 0
+    if not gathered:
+        log.info("There is no prefetched source archive.")
+
+    cachi2_env = f"{cachi2_dir}/cachi2.env"
+    if os.path.exists(cachi2_env):
+        prepared_env_dir = create_dir(work_dir, "cachi2_env")
+        src = cachi2_env
+        dest = f"{prepared_env_dir}/cachi2.env"
+        log.debug("copy cachi2 env file %s to %s", src, dest)
+        shutil.copy(src, dest)
+        sib_dirs.extra_src_dirs.append(prepared_env_dir)
+
+    return gathered
+
+
+def make_source_archive(
+    work_dir: str, source_dir: str, sib_dirs: SourceImageBuildDirectories
+) -> None:
+    log = logging.getLogger("build-source.source-archive")
+    source_archive_dir = create_dir(work_dir, "source_archive")
+    repo_info = get_repo_info(source_dir)
+    name_sha = f"{repo_info['name']}-{repo_info['last_commit_sha']}"
+    output_archive = f"{source_archive_dir}/{name_sha}.tar.gz"
+    git_cmd = ["git", "archive", "--prefix", name_sha + "/", "--output", output_archive, "HEAD"]
```

Comment:
So this takes the clean state at the latest commit and throws away the modifications that we do, right? Like the `sed` commands in the buildah task, `cachi2 inject-files` etc.

I guess that's what we want?

Reply:
Another thing - the archive also excludes the .git/ directory, so the source container doesn't include the git history.

That's probably fine for the first implementation, but could be another issue if we're going to use source containers as the only source for re-builds (https://github.com/redhat-appstudio/build-definitions/pull/621#discussion_r1366959493)

Reply:
We can change that if/when needed, although there will be a slight issue that some existing source containers will not be usable for re-builds